packages/gnat/confinement/confinement.py

```diff
@@ -0,0 +1,167 @@
+        huc ([type]): [description]
+
+    Raises:
+        Exception: [description]
+        Exception: [description]
+        Exception: [description]
+        Exception: [description]
+
+    Returns:
+        [type]: [description]
+    """
+
+    log = Logger("Confinement")
+    log.info(f'Confinement v.{0.0}')  # .format(cfg.version))
+
+    try:
+        int(huc)
+    except ValueError:
+        raise Exception('Invalid HUC identifier "{}". Must be an integer'.format(huc))
+
+    if not (len(huc) == 4 or len(huc) == 8):
+        raise Exception('Invalid HUC identifier. Must be four digit integer')
+
+    safe_makedirs(output_folder)
+    out_gpkg = os.path.join(output_folder, "Confinement.gpkg")
+    if os.path.exists(out_gpkg):
+        os.remove(out_gpkg)
+    log.info(f"Preparing output geopackage: {out_gpkg}")
+    driver_gpkg = ogr.GetDriverByName("GPKG")
+    driver_gpkg.CreateDataSource(out_gpkg)
+
+    # Generate confining margins
+    log.info('Generating Confined Margins')
+    confining_margins, floodplain_pockets = generate_confining_margins(active_channel_polygon, confining_polygon, out_gpkg)
+
+    return
+
+
+def generate_confining_margins(active_channel_polygon, confining_polygon, output_gpkg, type="Unspecified"):
+    """[summary]
+
+    Args:
+        active_channel_polygon (str): featureclass of active channel
+        confining_margin_polygon (str): featureclass of confinig margins
+        confinement_type (str): type of confinement
+
+    Returns:
+        geometry: confining margins polylines
+        geometry: floodplain pockets polygons
+    """
+
+    # Load geoms
+    driver = ogr.GetDriverByName("ESRI Shapefile")  # GPKG
+    data_active_channel = driver.Open(active_channel_polygon, 0)
+    lyr_active_channel = data_active_channel.GetLayer()
+    data_confining_polygon = driver.Open(confining_polygon, 0)
+    lyr_confining_polygon = data_confining_polygon.GetLayer()
+    srs = lyr_active_channel.GetSpatialRef()
+    geom_active_channel_polygon = unary_union([wkb_load(feat.GetGeometryRef().ExportToWkb()) for feat in lyr_active_channel])
+    geom_confining_polygon = unary_union([wkb_load(feat.GetGeometryRef().ExportToWkb()) for feat in lyr_confining_polygon])
+
+    geom_confined_area = geom_active_channel_polygon.difference(geom_confining_polygon)
+    geom_confining_margins = geom_confined_area.boundary.intersection(geom_confining_polygon.boundary)
+    geom_floodplain_pockets = geom_confining_polygon.difference(geom_active_channel_polygon)
+
+    # TODO : clean/test outputs?
+
+    # Save Outputs to Geopackage
+
+    out_driver = ogr.GetDriverByName("GPKG")
+    data_out = out_driver.Open(output_gpkg, 1)
+    lyr_out_confining_margins = data_out.CreateLayer('ConfiningMargins', srs, geom_type=ogr.wkbLineString)
+    feature_def = lyr_out_confining_margins.GetLayerDefn()
+    feature = ogr.Feature(feature_def)
+    feature.SetGeometry(ogr.CreateGeometryFromJson(json.dumps(mapping(geom_confining_margins))))
```

Comment:
Take a look at this:

https://github.com/Riverscapes/riverscapes-tools/blob/master/lib/commons/rscommons/shapefile.py#L876

Going to and from wkb is less memory and less steps.